// hardhat-resolc/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArtifactFromContractOutput = getArtifactFromContractOutput;
exports.getVersionComponents = getVersionComponents;
exports.updateDefaultCompilerConfig = updateDefaultCompilerConfig;
exports.pluralize = pluralize;
exports.extractCommands = extractCommands;
exports.extractImports = extractImports;
exports.mapImports = mapImports;
exports.orderSources = orderSources;
exports.deepUpdate = deepUpdate;
const constants_1 = require("hardhat/internal/constants");
const constants_2 = require("./constants");
const chalk_1 = __importDefault(require("chalk"));
const errors_1 = require("./errors");
function getArtifactFromContractOutput(sourceName, contractName, contractOutput) {
    const evmBytecode = contractOutput.evm?.bytecode;
    let bytecode = evmBytecode?.object ?? "";
    const evmDeployedBytecode = contractOutput.evm?.deployedBytecode;
    let deployedBytecode = evmDeployedBytecode?.object ?? "";
    const linkReferences = evmBytecode?.linkReferences ?? {};
    const deployedLinkReferences = evmDeployedBytecode?.linkReferences ?? {};
    return {
        _format: constants_1.ARTIFACT_FORMAT_VERSION,
        contractName,
        sourceName,
        abi: contractOutput.abi,
        bytecode,
        deployedBytecode,
        linkReferences,
        deployedLinkReferences,
    };
}
function getVersionComponents(version) {
    const versionComponents = version.split('.');
    return [parseInt(versionComponents[0], 10), parseInt(versionComponents[1], 10), parseInt(versionComponents[2], 10)];
}
function updateDefaultCompilerConfig(solcConfigData, resolc) {
    const compiler = solcConfigData.compiler;
    const settings = compiler.settings || {};
    compiler.settings = { ...settings, optimizer: { ...resolc.settings?.optimizer }, evmVersion: resolc.settings?.evmVersion };
    const forceEVMLA = resolc.settings?.forceEVMLA && resolc.compilerSource === 'binary';
    resolc.settings.forceEVMLA = forceEVMLA;
    const [major, minor] = getVersionComponents(compiler.version);
    if (major === 0 && minor < 7 && resolc.compilerSource === 'binary') {
        console.warn(chalk_1.default.blue(constants_2.COMPILER_RESOLC_NEED_EVM_CODEGEN));
        compiler.settings.forceEVMLA = true;
    }
    delete compiler.settings.metadata;
}
function pluralize(n, singular, plural) {
    if (n === 1) {
        return singular;
    }
    if (plural !== undefined) {
        return plural;
    }
    return `${singular}s`;
}
function extractYulCommands(config, commandArgs) {
    const settings = config.settings;
    commandArgs.push(`--yul`);
    if (settings.solcPath) {
        commandArgs.push(`--solc=${settings.solcPath}`);
    }
    ;
    if (settings.optimizer?.enabled) {
        commandArgs.push(`--optimization=${settings.optimizer.parameters || '3'}`);
    }
    ;
    if (settings.llvmVerifyEach) {
        commandArgs.push(`--llvm-verify-each`);
    }
    ;
    if (settings.llvmDebugLogging) {
        commandArgs.push(`--llvm-debug-logging`);
    }
    ;
    if (settings.metadataHash) {
        if (settings.metadataHash !== 'none') {
            throw new errors_1.ResolcPluginError(`--metadata-hash only supported value is 'none'.`);
        }
        ;
        commandArgs.push(`--metadata-hash=${settings.metadataHash}`);
    }
    ;
    if (settings.debugOutputDir) {
        commandArgs.push(`--debug-output-dir=${settings.debugOutputDir}`);
    }
    ;
    if (settings.emitDourceDebugInfo) {
        commandArgs.push(`-g`);
    }
    ;
    if (settings.asm) {
        commandArgs.push(`--asm`);
    }
    ;
    if (settings.bin) {
        commandArgs.push(`--bin`);
    }
    ;
    if (settings.outputDir) {
        commandArgs.push(`--output-dir=${settings.outputDir}`);
    }
    ;
    return commandArgs;
}
function extractLlvmIRCommands(config, commandArgs) {
    const settings = config.settings;
    commandArgs.push(`--llvm-ir`);
    if (settings.optimizer?.enabled) {
        commandArgs.push(`--optimization=${settings.optimizer.parameters || '3'}`);
    }
    ;
    if (settings.llvmVerifyEach) {
        commandArgs.push(`--llvm-verify-each`);
    }
    ;
    if (settings.llvmDebugLogging) {
        commandArgs.push(`--llvm-debug-logging`);
    }
    ;
    if (settings.metadataHash) {
        if (settings.metadataHash !== 'none') {
            throw new errors_1.ResolcPluginError(`--metadata-hash only supported value is 'none'.`);
        }
        ;
        commandArgs.push(`--metadata-hash=${settings.metadataHash}`);
    }
    ;
    if (settings.debugOutputDir) {
        commandArgs.push(`--debug-output-dir=${settings.debugOutputDir}`);
    }
    ;
    if (settings.emitDourceDebugInfo) {
        commandArgs.push(`-g`);
    }
    ;
    if (settings.asm) {
        commandArgs.push(`--asm`);
    }
    ;
    if (settings.bin) {
        commandArgs.push(`--bin`);
    }
    ;
    if (settings.outputDir) {
        commandArgs.push(`--output-dir=${settings.outputDir}`);
    }
    ;
    return commandArgs;
}
function extractStandardJSONCommands(config, commandArgs) {
    const settings = config.settings;
    commandArgs.push(`--standard-json`);
    if (settings.solcPath) {
        commandArgs.push(`--solc=${settings.solcPath}`);
    }
    ;
    if (settings.detectMissingLibraries) {
        commandArgs.push(`--detect-missing-libraries`);
    }
    ;
    if (settings.forceEVMLA) {
        commandArgs.push(`--force-evmla`);
    }
    ;
    if (settings.basePath) {
        commandArgs.push(`--base-path=${settings.basePath}`);
    }
    ;
    if (settings.includePaths) {
        commandArgs.push(`--include-paths=${settings.includePaths}`);
    }
    ;
    if (settings.allowPaths) {
        if (!settings.basePath) {
            throw new errors_1.ResolcPluginError(`--allow-paths option is only available when --base-path has a non-empty value.`);
        }
        commandArgs.push(`--allow-paths=${settings.allowPaths}`);
    }
    ;
    if (settings.debugOutputDir) {
        commandArgs.push(`--debug-output-dir=${settings.debugOutputDir}`);
    }
    ;
    if (settings.emitDourceDebugInfo) {
        commandArgs.push(`-g`);
    }
    ;
    if (settings.asm) {
        commandArgs.push(`--asm`);
    }
    ;
    if (settings.bin) {
        commandArgs.push(`--bin`);
    }
    ;
    if (settings.outputDir) {
        commandArgs.push(`--output-dir=${settings.outputDir}`);
    }
    ;
    return commandArgs;
}
function extractCombinedJSONCommands(config, commandArgs) {
    const settings = config.settings;
    commandArgs.push(`--combined-json=${settings.combinedJson}`);
    if (settings.libraries) {
        commandArgs.push(`-l=${settings.libraries}`);
    }
    ;
    if (settings.solcPath) {
        commandArgs.push(`--solc=${settings.solcPath}`);
    }
    ;
    if (settings.evmVersion) {
        commandArgs.push(`--evm-version=${settings.evmVersion}`);
    }
    ;
    if (settings.disableSolcOptimizer) {
        commandArgs.push(`--disable-solc-optimizer`);
    }
    ;
    if (settings.optimizer?.enabled) {
        commandArgs.push(`--optimization=${settings.optimizer.parameters || '3'}`);
    }
    ;
    if (settings.forceEVMLA) {
        commandArgs.push(`--force-evmla`);
    }
    ;
    if (settings.metadataHash) {
        if (settings.metadataHash !== 'none') {
            throw new errors_1.ResolcPluginError(`--metadata-hash only supported value is 'none'.`);
        }
        ;
        commandArgs.push(`--metadata-hash=${settings.metadataHash}`);
    }
    ;
    if (settings.basePath) {
        commandArgs.push(`--base-path=${settings.basePath}`);
    }
    ;
    if (settings.includePaths) {
        commandArgs.push(`--include-paths=${settings.includePaths}`);
    }
    ;
    if (settings.allowPaths) {
        if (!settings.basePath) {
            throw new errors_1.ResolcPluginError(`--allow-paths option is only available when --base-path has a non-empty value.`);
        }
        commandArgs.push(`--allow-paths=${settings.allowPaths}`);
    }
    ;
    if (settings.suppressWarnings) {
        commandArgs.push(`--suppress-warnings=${settings.suppressWarnings.join(' ')}`);
    }
    ;
    if (settings.debugOutputDir) {
        commandArgs.push(`--debug-output-dir=${settings.debugOutputDir}`);
    }
    ;
    if (settings.emitDourceDebugInfo) {
        commandArgs.push(`-g`);
    }
    ;
    if (settings.overwrite) {
        commandArgs.push(`--overwrite`);
    }
    ;
    if (settings.asm) {
        commandArgs.push(`--asm`);
    }
    ;
    if (settings.bin) {
        commandArgs.push(`--bin`);
    }
    ;
    if (settings.outputDir) {
        commandArgs.push(`--output-dir=${settings.outputDir}`);
    }
    ;
    return commandArgs;
}
function extractRemainingCommands(config, commandArgs) {
    const settings = config.settings;
    if (settings.libraries) {
        commandArgs.push(`-l=${settings.libraries}`);
    }
    ;
    if (settings.solcPath) {
        commandArgs.push(`--solc=${settings.solcPath}`);
    }
    ;
    if (settings.evmVersion) {
        commandArgs.push(`--evm-version=${settings.evmVersion}`);
    }
    ;
    if (settings.disableSolcOptimizer) {
        commandArgs.push(`--disable-solc-optimizer`);
    }
    ;
    if (settings.optimizer?.enabled) {
        commandArgs.push(`--optimization=${settings.optimizer.parameters || '3'}`);
    }
    ;
    if (settings.forceEVMLA) {
        commandArgs.push(`--force-evmla`);
    }
    ;
    if (settings.metadataHash) {
        if (settings.metadataHash !== 'none') {
            throw new errors_1.ResolcPluginError(`--metadata-hash only supported value is 'none'.`);
        }
        ;
        commandArgs.push(`--metadata-hash=${settings.metadataHash}`);
    }
    ;
    if (settings.basePath) {
        commandArgs.push(`--base-path=${settings.basePath}`);
    }
    ;
    if (settings.includePaths) {
        commandArgs.push(`--include-paths=${settings.includePaths}`);
    }
    ;
    if (settings.allowPaths) {
        if (!settings.basePath) {
            throw new errors_1.ResolcPluginError(`--allow-paths option is only available when --base-path has a non-empty value.`);
        }
        commandArgs.push(`--allow-paths=${settings.allowPaths}`);
    }
    ;
    if (settings.suppressWarnings) {
        commandArgs.push(`--suppress-warnings=${settings.suppressWarnings.join(' ')}`);
    }
    ;
    if (settings.debugOutputDir) {
        commandArgs.push(`--debug-output-dir=${settings.debugOutputDir}`);
    }
    ;
    if (settings.emitDourceDebugInfo) {
        commandArgs.push(`-g`);
    }
    ;
    if (settings.asm) {
        commandArgs.push(`--asm`);
    }
    ;
    if (settings.bin) {
        commandArgs.push(`--bin`);
    }
    ;
    if (settings.outputDir) {
        commandArgs.push(`--output-dir=${settings.outputDir}`);
    }
    ;
    return commandArgs;
}
function extractCommands(config) {
    const commandArgs = [];
    if (config.settings?.yul) {
        return extractYulCommands(config, commandArgs);
    }
    else if (config.settings?.llvmIR) {
        return extractLlvmIRCommands(config, commandArgs);
    }
    else if (config.settings?.standardJson) {
        return extractStandardJSONCommands(config, commandArgs);
    }
    else if (config.settings?.combinedJson) {
        return extractCombinedJSONCommands(config, commandArgs);
    }
    else {
        return extractRemainingCommands(config, commandArgs);
    }
    ;
}
function extractImports(fileContent) {
    const importRegex = /import\s+(?:"([^"]+)"|'([^']+)'|(?:[^'"]+)\s+from\s+(?:"([^"]+)"|'([^']+)'))\s*;/g;
    let imports = [];
    let match;
    while ((match = importRegex.exec(fileContent)) !== null) {
        const importedPath = match[1] || match[2] || match[3] || match[4];
        if (importedPath) {
            imports.push(importedPath);
        }
    }
    return imports;
}
function mapImports(input) {
    const keys = Object.keys(input.sources);
    const map = new Map();
    for (let key of keys) {
        const importArray = extractImports(input.sources[key].content);
        map.set(key, importArray);
    }
    return map;
}
function orderSources(mapped) {
    let ordered = [];
    mapped.forEach((values, key) => {
        for (let value of values) {
            if (ordered.includes(value))
                continue;
            ordered.push(value);
        }
        if (ordered.includes(key))
            return;
        ordered.push(key);
    });
    return ordered;
}
function deepUpdate(a, b) {
    const keys = Object.keys(b.sources);
    const lastId = Object.keys(a.sources).length - 1;
    let nextIds = lastId + Object.keys(b.sources).length;
    if (Object.keys(a.sources).length === 0) {
        return b;
    }
    else {
        keys.forEach((key) => {
            a.contracts = Object.assign({}, a.contracts, { [key]: b.contracts[key] });
            a.sources = Object.assign({}, a.sources, { [key]: { id: nextIds, ast: b.sources[key].ast } });
            nextIds++;
        });
    }
    return a;
}
